cuda/matrix/csr_kernels.cu

```diff
@@ -307,23 +307,56 @@ void spmv(std::shared_ptr<const CudaExecutor> exec,
         if (cusparse::is_supported<ValueType, IndexType>::value) {
             // TODO: add implementation for int64 and multiple RHS
             auto handle = exec->get_cusparse_handle();
-            auto descr = cusparse::create_mat_descr();
             {
                 cusparse::pointer_mode_guard pm_guard(handle);
+                const auto alpha = one<ValueType>();
+                const auto beta = zero<ValueType>();
+                // TODO: add implementation for int64 and multiple RHS
+                if (b->get_stride() != 1 || c->get_stride() != 1)
+                    GKO_NOT_IMPLEMENTED;
+
+#if defined(CUDA_VERSION) && (CUDA_VERSION < 11000)
+                auto descr = cusparse::create_mat_descr();
                 auto row_ptrs = a->get_const_row_ptrs();
                 auto col_idxs = a->get_const_col_idxs();
-                auto alpha = one<ValueType>();
-                auto beta = zero<ValueType>();
-                if (b->get_stride() != 1 || c->get_stride() != 1) {
-                    GKO_NOT_IMPLEMENTED;
-                }
                 cusparse::spmv(handle, CUSPARSE_OPERATION_NON_TRANSPOSE,
                                a->get_size()[0], a->get_size()[1],
                                a->get_num_stored_elements(), &alpha, descr,
                                a->get_const_values(), row_ptrs, col_idxs,
                                b->get_const_values(), &beta, c->get_values());
+
+                cusparse::destroy(descr);
+#else  // CUDA_VERSION >= 11000
+                cusparseOperation_t trans = CUSPARSE_OPERATION_NON_TRANSPOSE;
+                cusparseSpMVAlg_t alg = CUSPARSE_CSRMV_ALG1;
+                auto row_ptrs =
+                    const_cast<IndexType *>(a->get_const_row_ptrs());
+                auto col_idxs =
+                    const_cast<IndexType *>(a->get_const_col_idxs());
+                auto values = const_cast<ValueType *>(a->get_const_values());
```

Comment:
Any idea why they take non const? That's a bit weird, do they want to change these somehow?

Reply:
They don't care about const correctness and want to allow themselves to reorder the columns if necessary, I would guess.